src/input/input_python.rs

```diff
@@ -347,6 +347,58 @@ impl<'a> Input<'a> for PyAny {
         }
     }
 
+    fn strict_decimal(&'a self, decimal_type: &Py<PyType>) -> ValResult<&'a PyAny> {
+        if self.is_instance(decimal_type.as_ref(self.py()))? {
+            Ok(self)
+        } else {
+            // FIXME make a known error
+            Err(ValError::new(
+                ErrorType::new_custom_error(PydanticCustomError::py_new(
+                    self.py(),
+                    "decimal_type".into(),
+                    "Input should be a valid Decimal instance or decimal string in JSON".into(),
+                    None,
+                )),
+                self,
+            ))
+        }
+    }
+
+    fn lax_decimal(&'a self, decimal_type: &Py<PyType>) -> ValResult<&'a PyAny> {
+        let py = self.py();
+
+        if self.is_instance(decimal_type.as_ref(self.py()))? {
+            return Ok(self);
+        }
+
+        let str = if let Ok(str) = self.downcast::<PyString>() {
+            str
+        } else if let Ok(float) = self.downcast::<PyFloat>() {
+            float.str()?
+        } else if let Ok(int) = self.downcast::<PyInt>() {
+            int.str()?
```

Comment:
given that we're just calling `.str()` and don't need the `PyFloat` / `PyInt`, it would probably be quicker to use `is_instance` - ideally we should use the `ObType` logic here so we only need to lookup the type pointer once.